src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -0,0 +1,470 @@
+{
+    if (! ctx.rules.enabled(featurePayChan,
+            ctx.app.config().features))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    auto const bal = ctx.tx[~sfBalance];
+    auto const amt = ctx.tx[~sfAmount];
+
+    if (ctx.tx[~sfSignature] &&
+        !(ctx.tx[~sfPublicKey] && bal))
+        return temMALFORMED;
+
+    if (bal && (!isXRP (*bal) || *bal <= beast::zero))
+        return temBAD_AMOUNT;
+
+    if (amt && (!isXRP (*amt) || *amt <= beast::zero))
+        return temBAD_AMOUNT;
+
+    if (bal && amt && *bal > *amt)
+        return tecNO_PERMISSION;
+
+    auto const flags = ctx.tx.getFlags ();
+    if ((flags & tfClose) && (flags & tfRenew))
+        return temMALFORMED;
+
+    return preflight2 (ctx);
+}
+
+TER
+PayChanClaim::doApply()
+{
+    Keylet const k (ltPAYCHAN, ctx_.tx[sfPayChannel]);
+    auto const slep = ctx_.view ().peek (k);
+    if (!slep)
+        return tecNO_TARGET;
+
+    AccountID const src = (*slep)[sfAccount];
+    AccountID const dst = (*slep)[sfDestination];
+    AccountID const txAccount = ctx_.tx[sfAccount];
+
+    auto const curExpiration = (*slep)[~sfExpiration];
+    {
+        auto const cancelAfter = (*slep)[~sfCancelAfter];
+        auto const closeTime =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (curExpiration && closeTime >= *curExpiration))
+            return closeChannel (
+                slep, ctx_.view (), k.key, ctx_.app.journal ("View"));
+    }
+
+    if (txAccount != src && txAccount != dst)
+        return tecNO_PERMISSION;
+
+    if (ctx_.tx[~sfBalance])
+    {
+        auto const chanBalance = slep->getFieldAmount (sfBalance).xrp ();
+        auto const chanFunds = slep->getFieldAmount (sfAmount).xrp ();
+        auto const reqBalance = ctx_.tx[sfBalance].xrp ();
+        auto const authAmt = [&] {
+            if (auto const& amt = ctx_.tx[~sfAmount])
+                return amt->xrp ();
+            return reqBalance;
+        }();
+
+        if (reqBalance > authAmt)
+            return tecNO_PERMISSION;
+
+        if (txAccount == dst && !ctx_.tx[~sfSignature])
+            return temBAD_SIGNATURE;
+
+        if (auto const& sig = ctx_.tx[~sfSignature])
+        {
+            // Check the signature
+            // The signature isn't needed if txAccount == src, but if it's
+            // present, check it
+
+            PublicKey const pk ((*slep)[sfPublicKey]);
+            if (ctx_.tx[sfPublicKey] != pk)
+                return temBAD_SIGNER;
+            Serializer msg;
+            std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
+            msg.addRaw (hdr, 4);
+            msg.add256 (k.key);
+            msg.add64 (authAmt.drops ());
+            if (!verify (pk, msg.slice (), *sig, /*canonical*/ true))
+                return temBAD_SIGNATURE;
+        }
```

Comment:
I moved the signature check into `preflight`. You suggested `preclaim`, but since the tx has the public key it can be checked without a ledger (the public key on the tx _is_ checked against the one stored in the ledger later.)